File: examples/order-python/app/application/services/order/place_order_service.py
from dataclasses import dataclass

from app.domain.order.model import Order
from app.domain.order.repository import OrderRepository
from app.domain.order.value import CustomerId, OrderId, OrderLine


@dataclass(frozen=True)
class PlaceOrderCommand:
    customer_id: str
    lines: list[tuple[str, int, int]]


class PlaceOrderService:
    def __init__(self, order_repository: OrderRepository) -> None:
        self._order_repository = order_repository

    def handle(self, command: PlaceOrderCommand) -> OrderId:
        order_id = OrderId.generate()
        customer_id = CustomerId(command.customer_id)
        lines = [
            OrderLine(product_id=product_id, quantity=quantity, unit_price=unit_price)
            for product_id, quantity, unit_price in command.lines
        ]
        order = Order.place(order_id, customer_id, lines)
        self._order_repository.save(order)
        order.pop_events()
        return order_id

File: examples/order-java/src/main/java/com/oopforge/example/order/application/service/PlaceOrderService.java
package com.oopforge.example.order.application.service;

import com.oopforge.example.order.application.provided.PlaceOrder;
import com.oopforge.example.order.application.required.OrderRepository;
import com.oopforge.example.order.domain.CustomerId;
import com.oopforge.example.order.domain.Order;
import com.oopforge.example.order.domain.OrderId;
import com.oopforge.example.order.domain.OrderLine;

import java.util.List;

public class PlaceOrderService implements PlaceOrder {

    private final OrderRepository orderRepository;

    public PlaceOrderService(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    @Override
    public OrderId handle(PlaceOrderCommand command) {
        OrderId orderId = OrderId.generate();
        CustomerId customerId = new CustomerId(command.customerId());
        List<OrderLine> lines = command.lines().stream()
                .map(line -> new OrderLine(line.productId(), line.quantity(), line.unitPrice()))
                .toList();

        Order order = Order.place(orderId, customerId, lines);
        orderRepository.save(order);
        order.popEvents();
        return orderId;
    }
}
